### PR TITLE
Add programmatic profiling APIs.

### PR DESCRIPTION
**Pull request by skye**
Also bumps the minimum jaxlib version, since these APIs depend on newly-exposed xla_client functionality.

### CODE REVIEW

File: docs/profiling.md [hunk 1]
```diff
@@ -18,7 +18,69 @@ Install specific nightly versions of TensorBoard, TensorBoard profiler, TensorFl
 pip install --upgrade tb-nightly==2.5.0a20201203 tbp-nightly==2.4.0a20201203 tf-nightly==2.5.0.dev20201203 tensorboard-plugin-wit==1.7.0
 ```
 
-### Usage
+### Programmatic capture
+
+You can instrument your code to capture a profiler trace via the
+{func}`jax.profiler.start_trace` and {func}`jax.profiler.stop_trace`
+methods. Call `start_trace` with the directory to write trace files to. This
```
Comment (jakevdp) [inline docs/profiling.md:25]: If you want, you can make this ``` {func}`~jax.profiler.start_trace` ``` which will render as **`start_trace()`** with a hyperlink to the source (the tilde removes the module path).

File: docs/profiling.md [hunk 2]
```diff
@@ -18,7 +18,69 @@ Install specific nightly versions of TensorBoard, TensorBoard profiler, TensorFl
...
+# Run the operations to be profiled
+key = jax.random.PRNGKey(0)
+x = jax.random.normal(key, (5000, 5000))
+y = x @ x
+y.block_until_ready()
+
+jax.profiler.stop_trace()
+```
+
+Note the {func}`block_until_ready` call. We use this to make sure on-device
```
Comment (jakevdp) [inline docs/profiling.md:45]: I think this has to be something like ``` {meth}`~jax.interpreters.xla._DeviceArray.block_until_ready` ``` in order to generate the link to the method docs.

Comment (jakevdp) [reply docs/profiling.md:45]: ... though we may not even have `DeviceArray` methods in our docs

File: docs/profiling.md [hunk 3]
```diff
@@ -76,10 +138,12 @@ from a running program.
...
 
-1. By default, the events in the trace viewer are mostly low-level internal JAX
-   functions. You can add your own events and functions by using
-   {func}`jax.profiler.TraceContext` and {func}`jax.profiler.trace_function` in
-   your code and capturing a new profile.
+### Adding custom trace events
+
+By default, the events in the trace viewer are mostly low-level internal JAX
+functions. You can add your own events and functions by using
+{func}`jax.profiler.TraceContext` and {func}`jax.profiler.trace_function` in
```
Comment (jakevdp) [inline docs/profiling.md:145]: I think this should be ``` {class}`jax.profiler.TraceContext` ```

**Review by jakevdp — COMMENTED** [review]
A few comments on the docs... I didn't look closely at the implementation

File: jax/_src/profiler.py [hunk 4]
```diff
@@ -24,15 +26,87 @@ def start_server(port: int):
...
+  with _profile_state.lock:
+    if _profile_state.profile_session is None:
+      raise RuntimeError("No profile started")
+    _profile_state.profile_session.stop_and_export(_profile_state.log_dir)
+    _profile_state.profile_session = None
+    _profile_state.log_dir = None
+
+
+@contextmanager
+def take_trace(log_dir):
```
Comment (hawkinsp) [inline jax/_src/profiler.py:87]: I note that TF calls this `tf.profiler.Profile`. I think I like the form of the name more, but I find `Profile` misleading.

Perhaps `Trace`?

and maybe we should rename `TraceContext` to `TraceMe` or `TraceAnnotation`?

**Review by hawkinsp — COMMENTED** [review]
Seems fine to me.

File: jax/_src/profiler.py (hunk 4 above)
Comment (skye) [reply jax/_src/profiler.py]: As discussed offline, I went with the oh-so-tasteful `trace`.

I also renamed TraceContext --> TraceAnnotation and other similar renames.

**Review by hawkinsp — APPROVED** [review]

**Pull request merged**